### PR TITLE
[Robosaur] #14107 mongodb support

### PR DESCRIPTION
**Pull request by daniel-jeremia**
# Changes
- Added database folder
- Added MongoDB initialization with TypeORM
- Added TypeORM definition for `document_queue` and `process_record`

# How to Use
- Call `setConfigByJSONFile`
- Call `initDatabase`
- Use file name `*.entity.ts` to define entity, and also use TypeORM decorators
- Call `getRepository(entityName)`

# Update 1
- Added TypeORM definition for `labels` and `teams` 
- Added `base-payload` as a base class for team entities (e.g. `15` collection)

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+services:
+  mongo:
+    image: mongo
+    restart: always
+    ports:
+      - 27017:27017
+    environment:
+      MONGO_INITDB_ROOT_USERNAME: vulcan
+      MONGO_INITDB_ROOT_PASSWORD: vulcan
+      MONGO_INITDB_DATABASE: vulcan
```
Comment (tonoman3g) [inline docker-compose.yaml:12]: is this docker-compose.yml used for development only?

File: src/database/entities/data.entity.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import { Column, Entity, ObjectID, ObjectIdColumn } from 'typeorm';
+
+@Entity({ name: '15' })
+export class DataPayload {
```
Comment (tonoman3g) [inline src/database/entities/data.entity.ts:4]: can we make this as abstract class and create a new concrete team for team 15?

my thought, there will be multiple tables like this entity

File: src/database/entities/document_queue.entity.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { Column, Entity, ObjectID, ObjectIdColumn } from 'typeorm';
+import { DataPayload } from './data.entity';
+
+@Entity()
+export class DocumentQueueEntity {
```
Comment (tonoman3g) [inline src/database/entities/document_queue.entity.ts:5]: if there is no entity name, then will it use the class name? DocumentQueueEntity becomes like this? document_queue?

File: src/transformer/pcw-transformer/helper/project-settings.mapper.spec.ts [hunk 4]
```diff
@@ -10,7 +10,7 @@ describe('projectSettingsMapper', () => {
       enableEditSentence: true,
       hideLabelerNamesDuringReview: false,
       hideLabelsFromInactiveLabelSetDuringReview: false,
-      hideOriginalSentencesDuringReview: true,
+      hideOriginalSentencesDuringReview: false,
```
Comment (tonoman3g) [inline src/transformer/pcw-transformer/helper/project-settings.mapper.spec.ts:13]: i think ML twist changed this, why did u revert this?

File: sample/database/config.json [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{
+  "database": {
```
Comment (tonoman3g) [inline sample/database/config.json:2]: after this [PR](https://github.com/datasaur-ai/robosaur/pull/54) is merged, please move the database config example to the config_rex.json file

File: docker-compose.yaml (hunk 1 above)
Comment (daniel-jeremia) [reply docker-compose.yaml:12]: yes @tonoman3g

File: src/database/entities/document_queue.entity.ts (hunk 3 above)
Comment (daniel-jeremia) [reply src/database/entities/document_queue.entity.ts:5]: it becomes snake_case. e.g. document_queue_entity @tonoman3g

File: src/transformer/pcw-transformer/helper/project-settings.mapper.spec.ts (hunk 4 above)
Comment (daniel-jeremia) [reply src/transformer/pcw-transformer/helper/project-settings.mapper.spec.ts:13]: hmm.. i think ini karna aku keep up with master di branch nya @tonoman3g

File: sample/database/config.json (hunk 5 above)
Comment (daniel-jeremia) [reply sample/database/config.json:2]: got it

Comment (tonoman3g) [reply sample/database/config.json:2]: merged @daniel-jeremia

File: src/database/entities/document_queue.entity.ts (hunk 3 above)
Comment (tonoman3g) [reply src/database/entities/document_queue.entity.ts:5]: oh. is the table name document_queue_entity?

iirc the table should be document_queue @daniel-jeremia

File: docker-compose.yaml [hunk 6]
```diff
@@ -0,0 +1,22 @@
+version: '3.1'
+
+services:
+  mongo:
+    image: mongo
```
Comment (tonoman3g) [inline docker-compose.yaml:5]: can you confirm to yosua which version are we using right now?

@daniel-jeremia

File: src/database/entities/document_queue.entity.ts (hunk 3 above)
Comment (daniel-jeremia) [reply src/database/entities/document_queue.entity.ts:5]: ah i see, okay i'll explicitly write the entity name

File: docker-compose.yaml (hunk 6 above)
Comment (daniel-jeremia) [reply docker-compose.yaml:5]: 4.12

File: src/transformer/pcw-transformer/helper/project-settings.mapper.spec.ts (hunk 4 above)
Comment (daniel-jeremia) [reply src/transformer/pcw-transformer/helper/project-settings.mapper.spec.ts]: @tonoman3g kept up rex with master and its fixed now

File: docker-compose.yaml (hunk 6 above)
Comment (tonoman3g) [reply docker-compose.yaml:5]: please set the version here @daniel-jeremia

File: src/database/entities/base-payload.entity.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+export abstract class BasePayload {
+  @ObjectIdColumn()
+  id: ObjectID;
+
+  @Column()
+  filename: string;
+
+  @Column()
+  reading_result: Object;
```
Comment (daniel-jeremia) [inline src/database/entities/base-payload.entity.ts:11]: This 3 objects still waiting for Yosua for the type definition

File: src/database/entities/process_record.entity.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import { Column, Entity, ObjectID, ObjectIdColumn } from 'typeorm';
+import { DataPayload } from './data';
+
+@Entity({ name: 'process_record' })
+export class ProcessRecordEntity {
+  @ObjectIdColumn()
+  _id: ObjectID;
+
+  @Column({ type: 'json' })
+  data?: DataPayload;
```
Comment (tonoman3g) [inline src/database/entities/process_record.entity.ts:10]: are the DataPayload attributes nested inside `data` attribute? or are they flattened as ProcessRecordEntity attributes?

From Yosua's document, i think those fields are flattened
```
{
    _id: ObjectID('...'),
    id: "",
    filename: ""
    parsed_image_dir: ""
   ...
}
```

File: src/database/entities/teams.entity.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import { Column, Entity, ObjectID, ObjectIdColumn, PrimaryColumn } from 'typeorm';
+
```
Comment (tonoman3g) [inline src/database/entities/teams.entity.ts:2]: please add constant for 15

`export TEAM_SURATINSTRUKSI = 15`

File: src/database/entities/labels.entity.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  _id: ObjectID;
+
+  @PrimaryColumn()
+  id: number;
+
+  @Column()
+  name: string;
+
+  @Column()
+  code: string;
```
Comment (tonoman3g) [inline src/database/entities/labels.entity.ts:16]: I didn't find `code` field in the document 🤔 

eh wait, the doc seems wrong

File: src/database/entities/labels.entity.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+  @Column((type) => TeamsEntity)
+  team: TeamsEntity;
+
+  @Column({ default: false })
+  is_from_emoji: boolean;
```
Comment (tonoman3g) [inline src/database/entities/labels.entity.ts:22]: hmm, if we have a new channel, should we create a new field? maybe we should ask Yosua

@daniel-jeremia

File: src/database/entities/labels.entity.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+  @PrimaryColumn()
+  id: number;
+
+  @Column()
+  name: string;
+
+  @Column()
+  code: string;
+
+  @Column((type) => TeamsEntity)
```
Comment (tonoman3g) [inline src/database/entities/labels.entity.ts:18]: is this relationship feature from typeorm? or is the team data really stored here as nested object?

File: src/database/entities/base-payload.entity.ts [hunk 13]
```diff
@@ -0,0 +1,70 @@
+import { Column, ObjectID, ObjectIdColumn, PrimaryColumn } from 'typeorm';
+
+export class ReadingResult {
+  [id_page_index: string]: Array<any>;
+}
+
+export class DocumentData {
+  [id_page_index: string]: {
+    [label_name: string]: {
+      id: string;
```
Comment (tonoman3g) [inline src/database/entities/base-payload.entity.ts:10]: put comment here

// id format's is uuid

File: src/database/entities/base-payload.entity.ts [hunk 14]
```diff
@@ -0,0 +1,70 @@
+import { Column, ObjectID, ObjectIdColumn, PrimaryColumn } from 'typeorm';
+
+export class ReadingResult {
+  [id_page_index: string]: Array<any>;
```
Comment (tonoman3g) [inline src/database/entities/base-payload.entity.ts:4]: put comment here by giving the example of id_page_index

File: src/database/entities/base-payload.entity.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
+import { Column, ObjectID, ObjectIdColumn, PrimaryColumn } from 'typeorm';
+
```
Comment (tonoman3g) [inline src/database/entities/base-payload.entity.ts:2]: in google docs, there is this section

```
“reading_result”: {
    “<id_page-index>”: [
        [
            [
                xmin:integer,
                ymin:integer,
                xmax:integer,
                ymax:integer
            ],
            text:string,
            label:string,
            reading-confidence:float
        ],
        …
    ],
    …
}

```

@daniel-jeremia please check, i also summon you in the docs

File: src/database/entities/process_record.entity.ts (hunk 8 above)
Comment (daniel-jeremia) [reply src/database/entities/process_record.entity.ts:10]: image
I followed this diagram, which one do you think is right? @tonoman3g

File: src/database/entities/labels.entity.ts (hunk 10 above)
Comment (daniel-jeremia) [reply src/database/entities/labels.entity.ts:16]: image
This is in the diagram also @tonoman3g

File: src/database/entities/labels.entity.ts (hunk 12 above)
Comment (daniel-jeremia) [reply src/database/entities/labels.entity.ts:18]: This will create a subcollection reference in MongoDB

Comment (tonoman3g) [reply src/database/entities/labels.entity.ts:18]: will the foreign key be team_id? to field `_id` that which type `ObjectId`?

@daniel-jeremia

File: src/database/entities/labels.entity.ts (hunk 10 above)
Comment (tonoman3g) [reply src/database/entities/labels.entity.ts:16]: oh oke, so the docs is wrong ya @daniel-jeremia

File: src/database/entities/process_record.entity.ts (hunk 8 above)
Comment (tonoman3g) [reply src/database/entities/process_record.entity.ts:10]: we should ask yosua @daniel-jeremia

File: src/database/entities/labels.entity.ts (hunk 12 above)
Comment (daniel-jeremia) [reply src/database/entities/labels.entity.ts:18]: Speaking in terms of `RDBMS`, yes foreign key will be `team_id` which refers to `_id` of `team` collection @tonoman3g

File: src/database/entities/process_record.entity.ts (hunk 8 above)
Comment (tonoman3g) [reply src/database/entities/process_record.entity.ts:10]: so this is already correct ya

Comment (daniel-jeremia) [reply src/database/entities/process_record.entity.ts:10]: From Yosua: nested di dalam object data

File: src/database/entities/labels.entity.ts (hunk 12 above)
Comment (tonoman3g) [reply src/database/entities/labels.entity.ts:18]: we can ignore this comment since we will delete this file @daniel-jeremia

**Review by tonoman3g — APPROVED** [review]

**Comment by tonoman3g** [comment]
eh sorry, please remove labels.entity.ts ya @daniel-jeremia

**Comment by daniel-jeremia** [comment]
@tonoman3g done

**Pull request merged**